[bug] A panic can sometimes occur when parsing a ill-formed ACE binary data

A panic can sometimes occur when parsing a ill-formed ACE binary data:

```
panic: runtime error: slice bounds out of range [4:0]

goroutine 1 [running]:
github.com/TheManticoreProject/winacl/ace.(*AccessControlEntry).Unmarshal(0xc00016f3a8, {0xc0001a4b84, 0x44, 0x4c})
	/root/go/pkg/mod/github.com/!the!manticore!project/winacl@v1.2.10/ace/AccessControlEntry.go:45 +0x156a
github.com/TheManticoreProject/winacl/acl.(*DiscretionaryAccessControlList).Unmarshal(0xc00016f8c0, {0xc0001a4b7c, 0x4c, 0x54})
	/root/go/pkg/mod/github.com/!the!manticore!project/winacl@v1.2.10/acl/DiscretionaryAccessControlList.go:39 +0x165
github.com/TheManticoreProject/winacl/securitydescriptor.(*NtSecurityDescriptor).Unmarshal(0xc00016f720, {0xc0001a4b60?, 0x68?, 0x1?})
	/root/go/pkg/mod/github.com/!the!manticore!project/winacl@v1.2.10/securitydescriptor/NtSecurityDescriptor.go:73 +0x30b
github.com/TheManticoreProject/Delegations/core/mode_audit.AuditRessourceBasedConstrainedDelegations({0x7ffd4caa931e?, 0xf?}, 0x7ffd4caa9343?, 0xd?, 0x54?, 0x93?, 0xaa?)
	/workspace/core/mode_audit/RessourceBasedConstrainedDelegations.go:73 +0x750
main.main()
	/workspace/main.go:649 +0x935
```

This should return a parsing error not a panic

--- a/ace/AccessControlEntry_test.go
+++ b/ace/AccessControlEntry_test.go
@@ -70,3 +70,49 @@ func TestAccessControlEntry_Size(t *testing.T) {
 		}
 	}
 }
+
+// TestAccessControlEntry_Unmarshal_MalformedNoPanic verifies that
+// Unmarshalling truncated or otherwise malformed ACE binary data returns a
+// parse error instead of panicking. Regression test for issue #30.
+func TestAccessControlEntry_Unmarshal_MalformedNoPanic(t *testing.T) {
+	cases := []struct {
+		name string
+		hex  string
+	}{
+		// Header declares Size=4 (exactly the header) but the ACE type
+		// (ACCESS_ALLOWED = 0x00) expects a Mask and SID after it.
+		{"access_allowed_size4_no_body", "00000400"},
+		// Header Size=8: header + mask, but no SID bytes.
+		{"access_allowed_size8_no_sid", "00000800deadbeef"},
+		// Object ACE claiming Size=16 with truncated GUID region.
+		{"object_ace_truncated", "0500100000000000010000000000000000000000"},
+		// Header Size=5: 1 trailing byte — too short for a 4-byte mask.
+		{"access_allowed_size5_one_body_byte", "0000050000"},
+		// Minimum ACE-type byte only, no header size field.
+		{"ace_single_byte", "00"},
+		{"ace_two_bytes", "0000"},
+		{"ace_three_bytes", "000000"},
+		// ACL-level buffer truncation: a DACL iteration delegating here
+		// must not crash the process on malformed content.
+		{"access_denied_size4", "01000400"},
+		{"system_audit_size4", "02000400"},
+	}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			raw, err := hex.DecodeString(tc.hex)
+			if err != nil {
+				t.Fatalf("hex decode: %v", err)
+			}
+			defer func() {
+				if r := recover(); r != nil {
+					t.Fatalf("Unmarshal panicked on malformed input %q: %v", tc.hex, r)
+				}
+			}()
+			var a AccessControlEntry
+			_, err = a.Unmarshal(raw)
+			if err == nil {
+				t.Fatalf("expected an error for malformed input %q, got nil", tc.hex)
+			}
+		})
+	}
+}

--- a/ace/aceflags/AccessControlEntryFlags.go
+++ b/ace/aceflags/AccessControlEntryFlags.go
@@ -1,6 +1,9 @@
 package aceflags
 
-import "strings"
+import (
+	"fmt"
+	"strings"
+)
 
 // https://learn.microsoft.com/en-us/dotnet/api/system.security.accesscontrol.aceflags?view=net-8.0
 const (
@@ -76,6 +79,9 @@ var AccessControlEntryFlagToName = map[uint8]string{
 //     model and determines the permissions or behavior associated with the
 //     flag.
 func (aceflag *AccessControlEntryFlag) Unmarshal(marshalledData []byte) (int, error) {
+	if len(marshalledData) < 1 {
+		return 0, fmt.Errorf("AccessControlEntryFlag unmarshal requires at least 1 byte, got %d", len(marshalledData))
+	}
 	aceflag.RawValue = uint8(marshalledData[0])
 	aceflag.Values = []uint8{}
 	aceflag.Flags = []string{}

--- a/ace/aceflags/AccessControlEntryFlags_test.go
+++ b/ace/aceflags/AccessControlEntryFlags_test.go
@@ -1,0 +1,18 @@
+package aceflags_test
+
+import (
+	"testing"
+
+	"github.com/TheManticoreProject/winacl/ace/aceflags"
+)
+
+// TestAccessControlEntryFlag_Unmarshal_EmptyReturnsError is a regression test
+// for issue #30: parsers must return an error on truncated input instead of
+// panicking with "index out of range".
+func TestAccessControlEntryFlag_Unmarshal_EmptyReturnsError(t *testing.T) {
+	f := aceflags.AccessControlEntryFlag{}
+	_, err := f.Unmarshal([]byte{})
+	if err == nil {
+		t.Fatal("expected an error for empty input, got nil")
+	}
+}

--- a/ace/acetype/AccessControlEntryType.go
+++ b/ace/acetype/AccessControlEntryType.go
@@ -1,5 +1,7 @@
 package acetype
 
+import "fmt"
+
 const (
 	ACE_TYPE_ACCESS_ALLOWED                 uint8 = 0x00 // Access-allowed ACE that uses the ACCESS_ALLOWED_ACE (section 2.4.4.2) structure.
 	ACE_TYPE_ACCESS_DENIED                  uint8 = 0x01 // Access-denied ACE that uses the ACCESS_DENIED_ACE (section 2.4.4.4) structure.
@@ -77,6 +79,10 @@ var AccessControlEntryTypeValueToName = map[uint8]string{
 //     representing the ACE type. This data is typically defined by the Windows
 //     model and indicates the type of access control entry.
 func (acetype *AccessControlEntryType) Unmarshal(marshalledData []byte) (int, error) {
+	if len(marshalledData) < 1 {
+		return 0, fmt.Errorf("AccessControlEntryType unmarshal requires at least 1 byte, got %d", len(marshalledData))
+	}
+
 	// Set the value of the ACE type
 	acetype.Value = uint8(marshalledData[0])
 

--- a/ace/acetype/AccessControlEntryType_test.go
+++ b/ace/acetype/AccessControlEntryType_test.go
@@ -28,3 +28,14 @@ func TestAccessControlEntryType_Involution(t *testing.T) {
 		t.Errorf("Involution test failed: expected 0x%02x, got 0x%02x", originalType.Value, parsedType.Value)
 	}
 }
+
+// TestAccessControlEntryType_Unmarshal_EmptyReturnsError is a regression test
+// for issue #30: parsers must return an error on truncated input instead of
+// panicking with "index out of range".
+func TestAccessControlEntryType_Unmarshal_EmptyReturnsError(t *testing.T) {
+	a := acetype.AccessControlEntryType{}
+	_, err := a.Unmarshal([]byte{})
+	if err == nil {
+		t.Fatal("expected an error for empty input, got nil")
+	}
+}

--- a/ace/mask/AccessControlMask.go
+++ b/ace/mask/AccessControlMask.go
@@ -25,6 +25,10 @@ type AccessControlMask struct {
 // Unmarshal populates the AccessControlMask from raw byte data.
 // It extracts the RawValue and determines the corresponding flags and their names.
 func (acm *AccessControlMask) Unmarshal(marshalledData []byte) (int, error) {
+	if len(marshalledData) < 4 {
+		return 0, fmt.Errorf("AccessControlMask unmarshal requires at least 4 bytes, got %d", len(marshalledData))
+	}
+
 	// Store the raw bytes and set the size
 	acm.RawBytes = marshalledData
 	acm.RawBytesSize = 4

--- a/ace/mask/AccessControlMask_test.go
+++ b/ace/mask/AccessControlMask_test.go
@@ -28,3 +28,17 @@ func TestAccessControlMask_Involution(t *testing.T) {
 		t.Errorf("Involution test failed: expected 0x%08x, got 0x%08x", originalMask.RawValue, parsedMask.RawValue)
 	}
 }
+
+// TestAccessControlMask_Unmarshal_TruncatedReturnsError asserts that Unmarshal
+// returns a parse error instead of panicking when fewer than 4 bytes are given.
+// Regression test for issue #30.
+func TestAccessControlMask_Unmarshal_TruncatedReturnsError(t *testing.T) {
+	for _, n := range []int{0, 1, 2, 3} {
+		buf := make([]byte, n)
+		var m AccessControlMask
+		_, err := m.Unmarshal(buf)
+		if err == nil {
+			t.Errorf("Unmarshal(%d bytes) expected error, got nil", n)
+		}
+	}
+}

--- a/acl/revision/AccessControlListRevision.go
+++ b/acl/revision/AccessControlListRevision.go
@@ -1,5 +1,7 @@
 package revision
 
+import "fmt"
+
 const (
 	ACL_REVISION    = 0x02
 	ACL_REVISION_DS = 0x04
@@ -15,6 +17,9 @@ type AccessControlListRevision struct {
 // Parameters:
 //   - rawBytes ([]byte): The byte slice to parse.
 func (aclrev *AccessControlListRevision) Unmarshal(marshalledData []byte) (int, error) {
+	if len(marshalledData) < 1 {
+		return 0, fmt.Errorf("AccessControlListRevision unmarshal requires at least 1 byte, got %d", len(marshalledData))
+	}
 	aclrev.Value = uint8(marshalledData[0])
 
 	return 1, nil

--- a/acl/revision/AccessControlListRevision_test.go
+++ b/acl/revision/AccessControlListRevision_test.go
@@ -50,6 +50,17 @@ func TestAccessControlListRevisionUnmarshal(t *testing.T) {
 	}
 }
 
+// TestAccessControlListRevisionUnmarshal_EmptyReturnsError is a regression test
+// for issue #30: parsers must return an error on truncated input instead of
+// panicking with "index out of range".
+func TestAccessControlListRevisionUnmarshal_EmptyReturnsError(t *testing.T) {
+	aclrev := revision.AccessControlListRevision{}
+	_, err := aclrev.Unmarshal([]byte{})
+	if err == nil {
+		t.Fatal("expected an error for empty input, got nil")
+	}
+}
+
 func TestAccessControlListRevisionMarshal(t *testing.T) {
 	tests := []struct {
 		name     string

--- a/object/flags/AccessControlObjectTypeFlags.go
+++ b/object/flags/AccessControlObjectTypeFlags.go
@@ -1,6 +1,9 @@
 package flags
 
-import "encoding/binary"
+import (
+	"encoding/binary"
+	"fmt"
+)
 
 // A set of bit flags that indicate whether the ObjectType and InheritedObjectType members are present. This parameter can be one or more of the following values.
 //
@@ -24,6 +27,9 @@ type AccessControlObjectTypeFlags struct {
 // Attributes:
 //   - RawBytes ([]byte): The byte slice representing the AccessControlObjectTypeFlags.
 func (acotype *AccessControlObjectTypeFlags) Unmarshal(rawBytes []byte) (int, error) {
+	if len(rawBytes) < 4 {
+		return 0, fmt.Errorf("AccessControlObjectTypeFlags unmarshal requires at least 4 bytes, got %d", len(rawBytes))
+	}
 	acotype.Value = binary.LittleEndian.Uint32(rawBytes[0:4])
 	return 4, nil
 }

--- a/object/flags/AccessControlObjectTypeFlags_test.go
+++ b/object/flags/AccessControlObjectTypeFlags_test.go
@@ -1,0 +1,21 @@
+package flags_test
+
+import (
+	"testing"
+
+	"github.com/TheManticoreProject/winacl/object/flags"
+)
+
+// TestAccessControlObjectTypeFlags_Unmarshal_TruncatedReturnsError asserts that
+// Unmarshal returns a parse error instead of panicking when fewer than 4 bytes
+// are provided. Regression test for issue #30.
+func TestAccessControlObjectTypeFlags_Unmarshal_TruncatedReturnsError(t *testing.T) {
+	for _, n := range []int{0, 1, 2, 3} {
+		buf := make([]byte, n)
+		var f flags.AccessControlObjectTypeFlags
+		_, err := f.Unmarshal(buf)
+		if err == nil {
+			t.Errorf("Unmarshal(%d bytes) expected error, got nil", n)
+		}
+	}
+}

--- a/securitydescriptor/NtSecurityDescriptor_fuzz_test.go
+++ b/securitydescriptor/NtSecurityDescriptor_fuzz_test.go
@@ -1,0 +1,55 @@
+package securitydescriptor_test
+
+import (
+	"testing"
+
+	"github.com/TheManticoreProject/winacl/securitydescriptor"
+)
+
+// FuzzNtSecurityDescriptor_Unmarshal asserts that Unmarshal never panics on
+// arbitrary input. A parse error is an acceptable outcome; a panic is not.
+// Regression test for issue #30.
+func FuzzNtSecurityDescriptor_Unmarshal(f *testing.F) {
+	seeds := [][]byte{
+		nil,
+		{},
+		{0x01},
+		{0x01, 0x00, 0x14, 0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x14, 0x00, 0x00, 0x00},
+	}
+	for _, s := range seeds {
+		f.Add(s)
+	}
+	f.Fuzz(func(t *testing.T, data []byte) {
+		defer func() {
+			if r := recover(); r != nil {
+				t.Fatalf("Unmarshal panicked on input len=%d: %v", len(data), r)
+			}
+		}()
+		ntsd := securitydescriptor.NtSecurityDescriptor{}
+		_, _ = ntsd.Unmarshal(data)
+	})
+}
+
+// FuzzNtSecurityDescriptor_FromSDDLString asserts that FromSDDLString never
+// panics on arbitrary string input.
+func FuzzNtSecurityDescriptor_FromSDDLString(f *testing.F) {
+	seeds := []string{
+		"",
+		"O:BAG:SYD:(A;;GA;;;WD)",
+		"D:(",
+		"S:(AU;SAFA;GA;;;WD",
+		"o:bag:syd:(a;;ga;;;wd)",
+	}
+	for _, s := range seeds {
+		f.Add(s)
+	}
+	f.Fuzz(func(t *testing.T, data string) {
+		defer func() {
+			if r := recover(); r != nil {
+				t.Fatalf("FromSDDLString panicked on input %q: %v", data, r)
+			}
+		}()
+		ntsd := securitydescriptor.NtSecurityDescriptor{}
+		_, _ = ntsd.FromSDDLString(data)
+	})
+}

--- a/securitydescriptor/control/NtSecurityDescriptorControl.go
+++ b/securitydescriptor/control/NtSecurityDescriptorControl.go
@@ -2,6 +2,7 @@ package control
 
 import (
 	"encoding/binary"
+	"fmt"
 )
 
 // NtSecurityDescriptorControl represents the control flags for a NT Security Descriptor.
@@ -79,6 +80,9 @@ var NtSecurityDescriptorControlValueToShortName = map[uint16]string{
 // Parameters:
 //   - rawValue (uint16): The raw value to be parsed, representing the control flags as a bitmask.
 func (nsdc *NtSecurityDescriptorControl) Unmarshal(rawValue []byte) (int, error) {
+	if len(rawValue) < 2 {
+		return 0, fmt.Errorf("NtSecurityDescriptorControl unmarshal requires at least 2 bytes, got %d", len(rawValue))
+	}
 	nsdc.RawValue = binary.LittleEndian.Uint16(rawValue)
 	nsdc.Values = []uint16{}
 	nsdc.Flags = []string{}

--- a/securitydescriptor/control/NtSecurityDescriptorControl_test.go
+++ b/securitydescriptor/control/NtSecurityDescriptorControl_test.go
@@ -78,3 +78,17 @@ func TestNtSecurityDescriptorControl_Marshal(t *testing.T) {
 		t.Errorf("Expected deserialized value to be 0x%04x, but got 0x%04x", uintValue, deserializedValue)
 	}
 }
+
+// TestNtSecurityDescriptorControl_Unmarshal_TruncatedReturnsError is a
+// regression test for issue #30: parsers must return a parse error on
+// truncated input instead of panicking.
+func TestNtSecurityDescriptorControl_Unmarshal_TruncatedReturnsError(t *testing.T) {
+	for _, n := range []int{0, 1} {
+		buf := make([]byte, n)
+		c := &control.NtSecurityDescriptorControl{}
+		_, err := c.Unmarshal(buf)
+		if err == nil {
+			t.Errorf("Unmarshal(%d bytes) expected error, got nil", n)
+		}
+	}
+}

--- a/sid/authority/SecurityIdentifierAuthority.go
+++ b/sid/authority/SecurityIdentifierAuthority.go
@@ -1,6 +1,9 @@
 package authority
 
-import "encoding/binary"
+import (
+	"encoding/binary"
+	"fmt"
+)
 
 // SID authority constants define the various authorities used in Security Identifiers (SIDs),
 // represented as hexadecimal values.
@@ -63,6 +66,10 @@ type SecurityIdentifierAuthority struct {
 //     If found, assigns the retrieved name to the `Name` field.
 //     If not found, assigns a default value of "?" to `Name`.
 func (sia *SecurityIdentifierAuthority) Unmarshal(marshalledData []byte) (int, error) {
+	if len(marshalledData) < 6 {
+		return 0, fmt.Errorf("SecurityIdentifierAuthority unmarshal requires at least 6 bytes, got %d", len(marshalledData))
+	}
+
 	sia.Value = 0
 	sia.Value += uint64(binary.BigEndian.Uint16(marshalledData[0:2])) << 32
 	sia.Value += uint64(binary.BigEndian.Uint16(marshalledData[2:4])) << 16

--- a/sid/authority/SecurityIdentifierAuthority_test.go
+++ b/sid/authority/SecurityIdentifierAuthority_test.go
@@ -25,3 +25,17 @@ func Test_SIDAuthorityValueToName(t *testing.T) {
 		}
 	}
 }
+
+// Test_SecurityIdentifierAuthority_Unmarshal_TruncatedReturnsError asserts that
+// Unmarshal returns a parse error instead of panicking when fewer than 6 bytes
+// are provided. Regression test for issue #30.
+func Test_SecurityIdentifierAuthority_Unmarshal_TruncatedReturnsError(t *testing.T) {
+	for _, n := range []int{0, 1, 2, 3, 4, 5} {
+		buf := make([]byte, n)
+		var a authority.SecurityIdentifierAuthority
+		_, err := a.Unmarshal(buf)
+		if err == nil {
+			t.Errorf("Unmarshal(%d bytes) expected error, got nil", n)
+		}
+	}
+}

